### ISSUE TITLE
Data race in linux `uv__hrtime()` when setting `fast_clock_id`

### ISSUE THREAD

**Issue by beauby** [libuv/libuv#2884]
* **Version**: 1.35.0
* **Platform**:  Linux ff1cd0dbd074 4.15.0-1067-aws #71-Ubuntu SMP Thu Apr 30 13:51:43 UTC 2020 x86_64 Linux

The lazy initialization of `fast_clock_id` can race when two uv loops are spawned from the same process. The race (introduced in 21c37a7db83e5e06cb28f61112325751fb41661c) seems harmless as the initialization seems deterministic, but it makes TSAN complain.

**Comment by bnoordhuis** [comment]
It's a pattern that's used throughout libuv: feature-detect X and record the result in a variable. 

I don't have an immediate suggestion on how to silence TSAN. If this were a C11 project, we'd load and store such variables with `memory_order_relaxed`, but libuv is C89.

We can probably switch to `__atomic_load_n(&fast_clock_id, __ATOMIC_RELAXED)` and `__atomic_store_n(&fast_clock_id, value, __ATOMIC_RELAXED)` (but macro-wrapped to paper over compiler differences), that should give TSAN enough compile-time information.

**Comment by bnoordhuis** [comment]
@beauby Can you check if #2886 fixes the issue for you?

**Comment by beauby** [comment]
Thanks for the quick fix! I just tested it and it works.

**Comment by beauby** [comment]
Either the PR was modified before it was merged, or I something went wrong when I tried it, but I am still hitting this on master. The following change seems to fix it. 

```
--- a/src/unix/linux-core.c
+++ b/src/unix/linux-core.c
@@ -483,18 +483,18 @@ uint64_t uv__hrtime(uv_clocktype_t type) {
   /* TODO(bnoordhuis) Use CLOCK_MONOTONIC_COARSE for UV_CLOCK_PRECISE
    * when it has microsecond granularity or better (unlikely).
    */
-  if (type == UV_CLOCK_FAST && fast_clock_id == -1) {
+  if (type == UV_CLOCK_FAST && uv__load_relaxed(&fast_clock_id) == -1) {
     if (clock_getres(CLOCK_MONOTONIC_COARSE, &t) == 0 &&
         t.tv_nsec <= 1 * 1000 * 1000) {
-      fast_clock_id = CLOCK_MONOTONIC_COARSE;
+      uv__store_relaxed(&fast_clock_id, CLOCK_MONOTONIC_COARSE);
     } else {
-      fast_clock_id = CLOCK_MONOTONIC;
+      uv__store_relaxed(&fast_clock_id, CLOCK_MONOTONIC);
     }
   }

   clock_id = CLOCK_MONOTONIC;
   if (type == UV_CLOCK_FAST)
-    clock_id = fast_clock_id;
+    clock_id = uv__load_relaxed(&fast_clock_id);

   if (clock_gettime(clock_id, &t))
     return 0;  /* Not really possible. */
--
2.17.1
```

**Comment by bnoordhuis** [comment]
Can you try the v1.x branch? That's what releases are cut from.

**Comment by beauby** [comment]
Sorry, I wrote "master" but I actually meant the v1.x branch.
Looking at #2886 it does not seem `uv__hrtime()` was modified.

**Comment by bnoordhuis** [comment]
Sorry, you're right; I dropped that part of the PR when I did the final rebase... Corrected in commit 10a9c25a.
